codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -0,0 +1,319 @@
+            Http2Stream stream = http2Handler.connection().requireStream(streamId);
+            final StreamInfo streamInfo = (StreamInfo) stream.getProperty(StreamInfo.class);
+            EventLoop childLoop = streamInfo.childChannel.eventLoop();
+            if (childLoop.inEventLoop()) {
+                childOnDataRead(streamInfo, data, endOfStream);
+            } else {
+                childLoop.execute(new Runnable() {
+                    @Override
+                    public void run() {
+                        childOnDataRead(streamInfo, data, endOfStream);
+                    }
+                });
+            }
+            return data.readableBytes() + padding;
+        }
+
+        void childOnDataRead(StreamInfo streamInfo, ByteBuf data, boolean endOfStream) {
+            fireChildReadAndRegister(streamInfo, data);
+            if (endOfStream) {
+                // TODO
+            }
+        }
+    }
+
+    static class StreamInfo {
+        final Http2StreamChannel childChannel;
+        /**
+         * True if stream is in {@link Http2MultiplexCodec#steamsToFireChildReadComplete}.
+         */
+        boolean inStreamsToFireChildReadComplete;
+
+        StreamInfo(Http2StreamChannel childChannel) {
+            this.childChannel = childChannel;
+        }
+    }
+
+    class Http2StreamChannel extends ChildChannel {
+        private final Http2Stream stream;
+
+        Http2StreamChannel(Http2Stream stream) {
+            super(Http2MultiplexCodec.this.ctx.channel());
+            this.stream = stream;
+        }
+
+        @Override
+        public void doClose() throws Exception {
+            EventLoop parentLoop = ctx.channel().eventLoop();
+            if (parentLoop.inEventLoop()) {
+                parentClose();
+            } else {
+                parentLoop.execute(new Runnable() {
+                    @Override
+                    public void run() {
+                        parentClose();
+                    }
+                });
+            }
+            super.doClose();
+        }
+
+        public void parentClose() {
+            try {
+                if (stream.state() != Http2Stream.State.CLOSED) {
+                    http2Handler.resetStream(ctx, stream.id(), Http2Error.CANCEL.code(),
+                        ctx.newPromise());
+                }
+            } catch (Throwable t) {
+                parentExceptionCaught(t);
+            }
+        }
+
+        @Override
+        public void parentWrite(Object msg) throws Exception {
+            // TODO: fix endOfStream
+            if (msg instanceof Http2Headers) {
+                http2Handler.encoder().writeHeaders(ctx,
+                    stream.id(), (Http2Headers) msg, 0, false, ctx.newPromise());
+            } else if (msg instanceof ByteBuf) {
+                http2Handler.encoder().writeData(ctx,
+                    stream.id(), (ByteBuf) msg, 0, true, ctx.newPromise());
+            }
```

Comment:
I think we still want `Http2FrameWriter` to do the actual writing, due to flow control.  We'll have to split data "frames" (I use the term loosely here, since they're not really frames until they hit the wire) and we wouldn't want to allocate a new wrapper object for each write (which may be fairly small in some cases). 

So the flow controller would take Headers/Data objects and write them using the `Http2FrameWriter`. Likewise, the rest of the codec will be changed to take the objects instead of having separate methods for each frame type.